src/Microsoft.ML.FastTree/Dataset/IntArray.cs

```diff
@@ -32,6 +34,38 @@ internal abstract class IntArray : IEnumerable<int>
         /// </summary>
         public abstract int Length { get; }
 
+        // We can test if the dll is present by calling into a non-existant method.
+        // If we get an EntryPointNotFoundException then the dll exists and we can use the native code.
+        // If we get a DllNotFoundException then the dll doesn't exist and we should use the managed fallbacks.
+        private const string NativePath = "FastTreeNative";
+        [DllImport(NativePath), SuppressUnmanagedCodeSecurity]
+        private static extern void NonExistantMethod();
+
+        /// <summary>
+        /// Lazy bool that checks if the FastTreeNative dll exists so we know if we should use the native code
+        /// or the managed fallbacks.
+        /// </summary>
+        public static Lazy<bool> UseFastTreeNative = new(() =>
+        {
+            try
+            {
+                NonExistantMethod();
+            }
+            // If the entry point isn't found then the dll exists and we should use the managed code.
+            catch(EntryPointNotFoundException)
```

Comment:
This seems like a hack. Can we just check the Current architecture, if it is x86 or x64, we call into the native code. Else, use the C# fallback.

[https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.processarchitecture](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.runtimeinformation.processarchitecture)

Reply:
I did it this way for 2 reasons.

1 - So that the managed fallback would work even on an x86 or x64 machine if desired. Do you think that situation will happen?

2 - The docs you linked say that call is supported starting in .NET Framework 4.7.1, but ML.NET supports .NET Framework 4.6.1. I still get mixed up about which things work as long as you build with a later version and which things require the later version at runtime. Would this still work if it was built with a version that supports it but runs on a version that doesn't?

Reply:
> So that the managed fallback would work even on an x86 or x64 machine if desired. Do you think that situation will happen?

It hasn't happened yet, so I don't think it is necessarily something we need to worry about. We don't do it in any other places we call into native code.

> The docs you linked say that call is supported starting in .NET Framework 4.7.1

That is a bit mis-leading. What that information is trying to convey is that the API is "inbox" on .NET Framework 4.7.1. If you look at the next row down you'll see the API supports .NET Standard `1.1` - `2.1`. And since ML.NET targets `netstandard2.0`, we can use this API.

Reply:
Good to know. I have switched it to use the code you suggested.